Modules/Filtering/Smoothing/include/itkFFTDiscreteGaussianImageFilter.hxx

```diff
@@ -0,0 +1,208 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkFFTDiscreteGaussianImageFilter_hxx
+#define itkFFTDiscreteGaussianImageFilter_hxx
+
+#include "itkFFTDiscreteGaussianImageFilter.h"
+
+#include "itkFFTPadImageFilter.h"
+#include "itkForwardFFTImageFilter.h"
+#include "itkInverseFFTImageFilter.h"
+#include "itkMultiplyImageFilter.h"
+
+#include "itkGaussianImageSource.h"
+#include "itkFFTConvolutionImageFilter.h"
+
+#include "itkNeighborhoodOperatorImageFilter.h"
+#include "itkGaussianOperator.h"
+#include "itkImageRegionIterator.h"
+#include "itkProgressAccumulator.h"
+#include "itkImageAlgorithm.h"
+#include "itkMacro.h"
+
+namespace itk
+{
+
+template <typename TInputImage, typename TOutputImage>
+void
+FFTDiscreteGaussianImageFilter<TInputImage, TOutputImage>::GenerateInputRequestedRegion()
+{
+  // call the superclass' implementation of this method. this should
+  // copy the output requested region to the input requested region
+  ImageToImageFilter::GenerateInputRequestedRegion();
+
+  // Get pointer to input
+  typename Superclass::InputImagePointer inputPtr = const_cast<TInputImage *>(this->GetInput());
+  if (inputPtr.IsNull())
+  {
+    return;
+  }
+
+  // get a copy of the input requested region (should equal the output
+  // requested region)
+  typename TInputImage::RegionType inputRequestedRegion;
+  inputRequestedRegion = inputPtr->GetRequestedRegion();
+
+  // pad the input requested region by the operator radius
+  typename TInputImage::SizeType radius;
+  radius.Fill(0);
+  for (size_t dim = 0; dim < ImageDimension; ++dim)
+  {
+    radius[dim] = (this->GetMaximumKernelWidth() - 1) / 2;
+  }
+  inputRequestedRegion.PadByRadius(radius);
+
+  // crop the input requested region at the input's largest possible region
+  if (inputRequestedRegion.Crop(inputPtr->GetLargestPossibleRegion()))
+  {
+    inputPtr->SetRequestedRegion(inputRequestedRegion);
+    return;
+  }
+  else
+  {
+    // Couldn't crop the region (requested region is outside the largest
+    // possible region).  Throw an exception.
+
+    // store what we tried to request (prior to trying to crop)
+    inputPtr->SetRequestedRegion(inputRequestedRegion);
+
+    // build an exception
```

Comment:
`itkExceptionMacro` is a more convenient way to raise an exception, e.g.: `itkExceptionMacro(<< "Requested region " << inputRequestedRegion << " is (at least partially) outside the largest possible region " << inputPtr->GetLargestPossibleRegion());`